src/Storages/StorageDistributed.cpp

```diff
@@ -1393,6 +1397,61 @@ std::vector<DistributedAsyncInsertDirectoryQueue::Status> StorageDistributed::ge
     return statuses;
 }
 
+Cluster::Addresses StorageDistributed::parseAddresses(const std::string & name) const
+{
+    Cluster::Addresses addresses;
+
+    const auto & cluster = getCluster();
+    const auto & shards_info = cluster->getShardsInfo();
+    const auto & shards_addresses = cluster->getShardsAddresses();
+
+    for (auto it = boost::make_split_iterator(name, boost::first_finder(",")); it != decltype(it){}; ++it)
+    {
+        const std::string & dirname = boost::copy_range<std::string>(*it);
+        Cluster::Address address = Cluster::Address::fromFullString(dirname);
+
+        /// Check new format shard{shard_index}_replica{replica_index}
+        /// (shard_index and replica_index starts from 1).
+        if (address.shard_index)
+        {
+            if (address.shard_index > shards_info.size())
+            {
+                LOG_ERROR(log, "No shard with shard_index={} ({})", address.shard_index, name);
+                continue;
+            }
+
+            const auto & replicas_addresses = shards_addresses[address.shard_index - 1];
+            size_t replicas = replicas_addresses.size();
+
+            if (dirname.ends_with("_all_replicas"))
+            {
+                for (size_t replica_index = 1; replica_index <= replicas; ++replica_index)
+                {
+                    Cluster::Address replica_address = replicas_addresses[replica_index - 1];
+                    replica_address.shard_index = address.shard_index;
+                    replica_address.replica_index = static_cast<UInt32>(replica_index);
+                    addresses.emplace_back(std::move(replica_address));
+                }
+                continue;
+            }
+
+            if (address.replica_index > replicas)
+            {
+                LOG_ERROR(log, "No shard with replica_index={} ({})", address.replica_index, name);
+                continue;
+            }
+
+            Cluster::Address replica_address = replicas_addresses[address.replica_index - 1];
+            replica_address.shard_index = address.shard_index;
+            replica_address.replica_index = address.replica_index;
+            addresses.emplace_back(std::move(replica_address));
```

Comment:
> AFAIR the only reason this code exists is that because `Address::operator==` does not have `shard_index`/`replica_index` and it always return `true` because everything else is empty in this case, why not simply add this two fields into comparator? This should be safe I guess... but kind of odd.

Another reason that comes to mind is that when host_name changes, the connection pool also needs to be updated.

> Another way is to remove handling Address with shard_index/replica_index completely from the DistributedAsyncInsertDirectoryQueue::createPool, let's do this instead then, to ensure that no more users left.

Do you mean deleting this code? https://github.com/ClickHouse/ClickHouse/blob/b7fa342e4cf6ce89fefdef2f8878977f8cd6a452/src/Storages/Distributed/DistributedAsyncInsertDirectoryQueue.cpp#L242-#L260